### PR TITLE
Adding autoscaling functionality

### PR DESCRIPTION
**Pull request by nhibberd**
- autoscaling types
- autoscaling commands
- optional aws resource tests
- updated doc on permissions / environment variables to control tests
- ec2 + autoscaling cabal builds

/cc @markhibberd @charleso

### CODE REVIEW

File: mismi-autoscaling/src/Mismi/Autoscaling/Error.hs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import           Control.Monad.Catch (Handler (..), throwM, catch)
+import           Control.Retry (RetryPolicyM, RetryStatus, recovering)
+
+import           Mismi.Amazonka (AWS, Error (..), ServiceError (..), errorCode, serviceCode, serviceStatus)
+
+import           Network.HTTP.Types.Status (status400)
+
+import           P
+
+retry :: RetryPolicyM AWS -> [Text] -> AWS a -> AWS a
```
Comment (charleso) [inline mismi-autoscaling/src/Mismi/Autoscaling/Error.hs]: Want to `newtype` the `Text`s here?

File: README.md [hunk 2]
```diff
@@ -6,34 +6,57 @@ mismi
...
+
+Resources that are used when running test are configurable via:
 
-Running amazonka `AWST` - [Mismi.Control.Amazonka](https://github.com/ambiata/mismi/blob/master/mismi-core/src/Mismi/Control/Amazonka.hs#L81)
 ```
-runAWSWithCreds :: Region -> AccessKey -> SecretKey -> Maybe SecurityToken -> Maybe UTCTime -> AWS a -> EitherT AWSError IO a
+`AWS_TEST_SECURITY_GROUP` - defaults to "ci.ci.node"
+`AWS_TEST_IMAGE_ID` - defaults to "ami-bc3611df"
+`AWS_TEST_IAM_ROLE` - defaults to "ci.ci.node"
+`AWS_TEST_BUCKET` - defaults to "ambiata-dev-view"
```
Comment (charleso) [inline README.md]: Nice 👍

File: mismi-autoscaling/src/Mismi/Autoscaling/Error.hs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+alreadyExists :: Text
+alreadyExists =
+  "AlreadyExists"
+
+throttling :: Text
+throttling =
+  "Throttling"
+
+handleServiceError :: (ServiceError -> Bool) -> a -> AWS a -> AWS a
```
Comment (charleso) [inline mismi-autoscaling/src/Mismi/Autoscaling/Error.hs]: Some of these look pretty generic, should they live in core?

File: mismi-s3/test/test-io.hs [hunk 4]
```diff
@@ -2,11 +2,13 @@ import           Disorder.Core.Main
...
 
 main :: IO ()
 main =
-  disorderMain [
-      Test.IO.Mismi.S3.Commands.tests
-    , Test.IO.Mismi.S3.Internal.tests
+  disorderMain =<< enableTests "AWS_TEST_S3" [
+      Test.IO.Mismi.S3.Internal.tests
+    ] [
+      Test.IO.Mismi.S3.Internal.tests
```
Comment (charleso) [inline mismi-s3/test/test-io.hs]: Would it make sense to always include the tests from the first list in the second? /notsure

Comment (nhibberd) [inline mismi-s3/test/test-io.hs]: I think it probably would, but thinking being precise about each list is better for the time being

File: mismi-autoscaling/src/Mismi/Autoscaling/Error.hs (hunk 3 above)
Comment (nhibberd) [inline mismi-autoscaling/src/Mismi/Autoscaling/Error.hs]: done

File: mismi-autoscaling/src/Mismi/Autoscaling/Error.hs (hunk 1 above)
Comment (nhibberd) [inline mismi-autoscaling/src/Mismi/Autoscaling/Error.hs]: rgr

**Comment by charleso** [comment]
👍

**Pull request merged**